src/java/org/apache/cassandra/cql3/Validation.java

```diff
@@ -64,4 +67,20 @@ public static void validateKey(TableMetadata metadata, ByteBuffer key)
             throw new InvalidRequestException(e.getMessage());
         }
     }
+
+    public static void checkConstraints(TableMetadata metadata, ByteBuffer key)
+    {
+        List<ColumnConstraint> partitionKeyConstraints = metadata.partitionKeyConstraints;
```

Comment:
@bbotella I believe this might be "cached" too. `metadata.hasPartitionKeyConstraints()` returning a boolean

Reply:
Actually, it is not necessary to resolve this every time. You can just do

````
            try
            {
                metadata.partitionKeyType.checkConstraints(key, metadata.partitionKeyConstraints);
            }
            catch (ConstraintViolationException e)
            {
                throw new InvalidRequestException(e.getMessage(), e);
            }
````

Because `checkConstraints` is again checking if it is empty. You are basically checking if it is empty twice.